tools/bin/mbedtls-rewrite-branch-style

```diff
@@ -0,0 +1,551 @@
+        #   * 'R{confidence}\t{old_name}\t{new_name}' for a renamed file
+        lines = self.get_git_lines(['show', '--format=', '--name-status',
+                                    revision])
+        filenames = [line.rsplit('\t', 1)[-1]
+                     for line in lines
+                     if line[0] != 'D'] # skip deleted files
+        return filenames
+
+    @staticmethod
+    def is_c_file(path: pathlib.PurePath) -> bool:
+        """Whether path is a C source file, based on its name and location."""
+        dir_chain = path.parts[:-1]
+        if path.suffix in {'.c', '.h'}:
+            return True
+        if dir_chain == ('tests', 'suites') and path.suffix == '.function':
+            return True
+        if dir_chain == ('scripts', 'data_files') and path.suffix == '.fmt':
+            return True
+        return False
+
+    def is_file_exempt(self, path: pathlib.PurePath) -> Optional[str]:
+        """Whether path is exempt from restyling.
+
+        Return the reason why path is exempt (a non-empty string),
+        or None if path is not exempt.
+        """
+        if path.parts[0] == '3rdparty':
+            return '3rdparty'
+        # Hard-coded list of files that are generated in 2.28.
+        # In development, generated files are not checked into Git.
+        if path.parts in self.GENERATED_FILES_2_28:
+            return 'generated'
+        return None
+
+    def filter_styled_files(self, filenames: List[str]) -> List[str]:
+        """Return filenames with non-restylable files removed."""
+        keep = []
+        for filename in filenames:
+            path = pathlib.PurePath(filename)
+            if not self.is_c_file(path):
+                continue
+            why_exempt = self.is_file_exempt(path)
+            if why_exempt:
+                self.info('Exempt from restyling ({}): {}', why_exempt, filename)
+                continue
+            self.info('Will restyle: {}', filename)
+            keep.append(filename)
+        return keep
+
+    def restyle_files(self, filenames: List[str]) -> None:
+        """Restyle the given files in place."""
+        if not filenames:
+            return
+        for num in range(2):
+            self.info('Pass #{} of uncrustify', num)
+            subprocess.check_call([self.UNCRUSTIFY_EXE,
+                                   '-c', self.CONFIG_FILE,
+                                   '--no-backup'] +
+                                  filenames)
+
+    def restyle_commit_onto_current(self, new_commit: str) -> None:
+        """Apply a restyled content of new_commit onto HEAD."""
+        self.info('Applying {} onto {}', new_commit, self.head_sha())
+        # Cherry-pick the commit to apply. Resolve all conflicts in
+        # favor of the new commit: we want the new commit's content,
+        # with the current commit's history.
+        self.run_git(['cherry-pick', '-Xtheirs', '--allow-empty', new_commit])
```

Comment:
It's working for me, unless I missed a problem with the result. My result is at https://github.com/Mbed-TLS/mbedtls/tree/features/new-code-style/test/new/6866 . What's wrong? Are you getting an error, a different result, or is the result wrong?

Reply:
Hmm that's interesting. The script passed for me but there are parts of `x509_crt.c` where [yours](https://github.com/Mbed-TLS/mbedtls/blob/features/new-code-style/test/new/6866/library/x509_crt.c#L2032) differs from [mine](https://github.com/davidhorstmann-arm/mbedtls/blob/restyled-extract-key-ids/library/x509_crt.c#L2061) (and mine is clearly wrong). Which version of Git are you using? Possibly there were some subtle changes to the default merging strategy or an unrecognized option or something.

Reply:
Ah, good to know! Git 2.25.1 on Ubuntu 20.04. I have a `~/.gitconfig` but I've run `tools/test/mbedtls-test-rewrite-branch-style` without it and it's passing.

Reply:
I'm using Git 2.38.1 so there might be a difference between the two. I'll see if I can install a newer version and test with that.

Reply:
The reason I'm using `git cherry-pick` is that it preserves metadata (author, commit message). But looking at the documentation of `-Xtheirs`, I think it says it isn't guaranteed to do what I want (which is to give files modified on both sides their content). I think a solution is to reset those files to the new commit's content, I should have a fix soon.

Reply:
Looks fine, except that I'm not sure what it will do in cases of deleted or renamed files - I imagine that `git reset` will complain that you're trying to reset files that don't exist.

Reply:
I haven't tested yet (can you find/construct a test case?) but shouldn't it work, since `reset` is only called on `changed_files`?

Reply:
Ah I hadn't looked at the details of `changed_files_in_commit()` - in that case it should work, but I'll make a test PR

Reply:
Test PR is Mbed-TLS/mbedtls#6874

Reply:
It doesn't need to be a PR, it just has to be a commit in the main repo, it can have a branch or tag called `features/new-code-style/test/old/*`

Reply:
Okay, I've put it in `features/new-code-style/test/old/rename-delete`.